lang/cs/Org.Apache.REEF.Common/Runtime/Evaluator/HeartBeatManager.cs

```diff
@@ -285,6 +302,12 @@ public void OnNext(Alarm value)
                     {
                         var driverConnection = _driverConnection.Get();
 
+                        if (driverConnection is MissingDriverConnection)
```

Comment:
I feel this code overcomplicates the logic. We know that `MissingDriverConnection.GetDriverInformation()` throws `NotImplementedException`, but theoretically it might not be the only implementation that does so. I'd replace this block with an extra `catch (NotImplementedException e)` clause at line 326 which would just re-throw the exception thrown by `MissingDriverConnection` (instead of swallowing it). And we don't need to log it here if we're throwing.